routing/dht/dht_bootstrap.go

```diff
@@ -0,0 +1,185 @@
+// of our implementation's robustness, we should lower this down to 8 or 4.
+//
+// Note there is also a tradeoff between the bootstrap period and the number
+// of queries. We could support a higher period with a smaller number of
+// queries
+const DefaultBootstrapQueries = 1
+
+// DefaultBootstrapPeriod specifies how often to periodically run bootstrap,
+// if the user does not specify a different number as an option.
+//
+// For now, this is set to 10 seconds, which is an aggressive period. We are
+// We are currently more interested in ensuring we have a properly formed
+// DHT than making sure our dht minimizes traffic. Once we are more certain
+// implementation's robustness, we should lower this down to 30s or 1m.
+//
+// Note there is also a tradeoff between the bootstrap period and the number
+// of queries. We could support a higher period with a smaller number of
+// queries
+const DefaultBootstrapPeriod = time.Duration(10 * time.Second)
+
+// DefaultBootstrapTimeout specifies how long to wait for a bootstrap query
+// to run.
+const DefaultBootstrapTimeout = time.Duration(10 * time.Second)
+
+// Bootstrap runs bootstrapping once, then calls SignalBootstrap with default
+// parameters: DefaultBootstrapQueries and DefaultBootstrapPeriod. This allows
+// the user to catch an error off the bat if the connections are faulty. It also
+// allows BootstrapOnSignal not to run bootstrap at the beginning, which is useful
+// for instrumenting it on tests, or delaying bootstrap until the network is online
+// and connected to at least a few nodes.
+//
+// Like PeriodicBootstrap, Bootstrap returns a process, so the user can stop it.
+func (dht *IpfsDHT) Bootstrap(ctx context.Context) (goprocess.Process, error) {
+
+	if err := dht.runBootstrap(ctx, DefaultBootstrapQueries); err != nil {
+		return nil, err
+	}
+
+	sig := time.Tick(DefaultBootstrapPeriod)
+	return dht.BootstrapOnSignal(DefaultBootstrapQueries, sig)
+}
+
+// SignalBootstrap ensures the dht routing table remains healthy as peers come and go.
+// it builds up a list of peers by requesting random peer IDs. The Bootstrap
+// process will run a number of queries each time, and run every time signal fires.
+// These parameters are configurable.
+//
+// SignalBootstrap returns a process, so the user can stop it.
+func (dht *IpfsDHT) BootstrapOnSignal(queries int, signal <-chan time.Time) (goprocess.Process, error) {
+	if queries <= 0 {
+		return nil, fmt.Errorf("invalid number of queries: %d", queries)
+	}
+
+	if signal == nil {
+		return nil, fmt.Errorf("invalid signal: %v", signal)
+	}
+
+	proc := periodicproc.Ticker(signal, func(worker goprocess.Process) {
+		// it would be useful to be able to send out signals of when we bootstrap, too...
+		// maybe this is a good case for whole module event pub/sub?
+
+		ctx := dht.Context()
+		if err := dht.runBootstrap(ctx, queries); err != nil {
+			log.Error(err)
+			// A bootstrapping error is important to notice but not fatal.
```

Comment:
We *could* also log an event here too, especially since, as you say, bootstrapping errors are something to take note of.